engine/runtime-integration-tests/src/test/java/org/enso/interpreter/runtime/RefTest.java

```diff
@@ -0,0 +1,124 @@
+  private static Value createRef;
+  private static Value getRef;
+
+  @BeforeClass
+  public static void initCtx() throws Exception {
+    ctx = ContextUtils.createDefaultContext();
+    ensoCtx = ContextUtils.leakContext(ctx);
+    var code =
+        """
+              import Standard.Base.Runtime.Ref.Ref
+
+              new_ref obj =
+                Ref.new obj
+
+              create_ref obj allow_gc =
+                Ref.new obj allow_gc
+
+              get_ref ref = ref.get
+              """;
+    var src = Source.newBuilder("enso", code, "gc.enso").build();
+    var gcEnso = ctx.eval(src);
+    newRef = gcEnso.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "new_ref");
+    createRef = gcEnso.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "create_ref");
+    getRef = gcEnso.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "get_ref");
+  }
+
+  @AfterClass
+  public static void closeCtx() throws Exception {
+    ctx.close();
+    ctx = null;
+  }
+
+  @Test
+  public void regularReference() throws Exception {
+    var obj = new Object();
+    var ref = newRef.execute(obj);
+
+    assertFalse("Value returned", ref.isNull());
+    assertEquals("Standard.Base.Runtime.Ref.Ref", ref.getMetaObject().getMetaQualifiedName());
+
+    var weakRef = new WeakReference<>(obj);
+    obj = null;
+
+    assertEquals("We get the object", weakRef.get(), getRef.execute(ref).asHostObject());
+
+    assertGC("Weak wasn't released", false, weakRef);
+    assertFalse("Value was not GCed", getRef.execute(ref).isNull());
+    assertEquals("We get the object", weakRef.get(), getRef.execute(ref).asHostObject());
+
+    ensoCtx.getReferencesManager().releaseAll();
+    assertEquals(
+        "releaseAll has no effect on regular reference",
+        weakRef.get(),
+        getRef.execute(ref).asHostObject());
+  }
+
+  @Test
+  public void garbageCollectableReference() throws Exception {
+    var obj = new Object();
+    var ref = createRef.execute(obj, true);
+
+    assertFalse("Value returned", ref.isNull());
+    assertEquals("Standard.Base.Runtime.Ref.Ref", ref.getMetaObject().getMetaQualifiedName());
+    assertEquals("We get the object", obj, getRef.execute(ref).asHostObject());
+
+    var weakRef = new WeakReference<>(obj);
+    obj = null;
+    assertGC("Weak reference can be GCed", true, weakRef);
+
+    assertTrue("Value was GCed", getRef.execute(ref).isNull());
+  }
+
+  @Test
+  public void explicitlyReclaimableReference() throws Exception {
+    var obj = new Object();
+    var ref = createRef.execute(obj, true);
+
+    assertFalse("Value returned", ref.isNull());
+    assertEquals("Standard.Base.Runtime.Ref.Ref", ref.getMetaObject().getMetaQualifiedName());
+    assertEquals("We get the object", obj, getRef.execute(ref).asHostObject());
+
+    ensoCtx.getReferencesManager().releaseAll();
+
+    assertTrue("Value was GCed", getRef.execute(ref).isNull());
+  }
+
+  private static void assertGC(String msg, boolean expectGC, Reference<?> ref) {
```

Comment:
Interesting test. Is it possible that it will cause transient issues on the CI? Is there any other (easy) way how to force GC, like via JMX?

Reply:
Here is the [inspiration](https://github.com/apache/netbeans/blob/master/harness/nbjunit/src/org/netbeans/junit/NbTestCase.java#L1436). Experience showed the original version to be quite reliable even when GCing `SoftReference` references.